src/testing/tree_construction.rs

```diff
@@ -206,22 +208,67 @@ impl Test {
 
     /// Run the parser and return the document and errors
     pub fn parse(&self, scripting_enabled: bool) -> Result<(DocumentHandle, Vec<ParseError>)> {
-        let document = DocumentBuilder::new_document();
+        // let mut is_fragment = false;
+        let mut context_node = None;
+        let document;
+
+        let is_fragment;
+
+        if let Some(fragment) = self.spec.document_fragment.clone() {
```

Comment:
This function now decides whether or not we are parsing a document or document-fragment based on the test specification. It will create a dummy main document with the context node, and then create the documen-fragment which gets filled by the parser.